### DISCUSSION TITLE
Why UT_MOD_DEPS sometimes is need also for fprime standard dependencies?

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1125] [Q&A]
On the [API.cmake](https://github.com/nasa/fprime/blob/devel/cmake/API.cmake) I found this comment about unit testing. I don't understand the difference between _Unit-Test Example_ and _Unit-Test Without GTest/TesterBase Example_. 
In the two example codes `UT_SOURCE_FILES` is set with the same files, based on the description of the example, should not be different(the second example without the .Ai file)?

```
# ### Unit-Test Example ###
#
# A standard unit test defines only UT_SOURCES. These sources have the test cpp files and the module
# Ai.xml of the module being tested. This is used to generate the GTest and TesterBase files from this
# Ai.xml. The other UT source files define the implementation of the test.
#
# ```
# set(UT_SOURCE_FILES
#   "${FPRIME_FRAMEWORK_PATH}/Svc/CmdDispatcher/CommandDispatcherComponentAi.xml"
#   "${CMAKE_CURRENT_LIST_DIR}/test/ut/CommandDispatcherTester.cpp"
#   "${CMAKE_CURRENT_LIST_DIR}/test/ut/CommandDispatcherImplTester.cpp"
# )
# register_fprime_ut()
# ```
#
# ### Unit-Test Without GTest/TesterBase Example ###
#
# Some unit tests run without the need for the autocoding the GTest and TesterBase files. This can be
# done without specifying the Ai.xml file. Most of the time, this style requires specifying some module
# dependencies.
#
# ```
# set(UT_SOURCE_FILES
#   "${FPRIME_FRAMEWORK_PATH}/Svc/CmdDispatcher/CommandDispatcherComponentAi.xml"
#   "${CMAKE_CURRENT_LIST_DIR}/test/ut/CommandDispatcherTester.cpp"
#   "${CMAKE_CURRENT_LIST_DIR}/test/ut/CommandDispatcherImplTester.cpp"
# )
# set(UT_MOD_DEPS
#   Os
# )
# register_fprime_ut()
# ```
```
NOTE: This question comes from the higher-level question of why for some of my modules I need `set(UT_MOD_DEPS Os)` and for other modules no.

**Comment by Joshua-Anderson** [comment]
This is a good question... I don't actually know why additional module dependencies must be specified if not generating the TesterBase class.  Maybe this style of unit test doesn't pull in the non-UT MOD_DEPS, but that seems unlikely

**Comment by LeStarch** [comment]
Module dependencies must be supplied when they cannot be detected automatically via the model.  The above documentation example specifies the `Os` dependency because often `Os` dependencies cannot be inferred from the model.  e.g. Using `Os::Mutex` in the implementation does not leave a trace in the Ai.xml file and thus the module dependency cannot be inferred.

Most of the time MOD_DEPS and UT_MOD_DEPS are not necessary and may be ommitted, however; the above documentation provides these as a reference.

**Comment by LeStarch** [comment]
To answer the other question about w/ and w/o GTest, this allows users to write UTs using the fprime UT framework without the need for GTest.  e.g. a user may wish to write UT style on-hardware tests but not compile against GTest.
